src/cosalib/metal.py

```diff
@@ -0,0 +1,316 @@
+"""
+Base Class for building Metal-type disks.
+"""
+
+# Copyright 2018-2020 Red Hat, Inc
+# Licensed under the new-BSD license (http://www.opensource.org/licenses/bsd-license.php)
+
+import distutils.util as dutil
+import logging as log
+import os
+import os.path
+import sys
+import yaml
+
+cosa_dir = os.path.dirname(os.path.abspath(__file__))
+sys.path.insert(0, f"{cosa_dir}/cosalib")
+sys.path.insert(0, cosa_dir)
+
+from cosalib.build import (
+    _Build,
+    BuildExistsError
+)
+from cosalib.ostree import _BuildOSTree
+from cosalib.cmdlib import (
+    get_basearch,
+    run_verbose,
+)
+
+
+# For padded, disk is actual plus 513M
+SIZE_PADDED = "513M"
+# For estimated sizes, inflate the actual rootfs by this percent
+SIZE_ESITMATED = 35
+# rootfs defaults to being 'xfs'
+DEFAULT_ROOTFS = "xfs"
+
+# VARIANTS are the type of disks that can build built
+# To define new variants add its definition below:
+VARIANTS = {
+    "qemu": {
+        "arch": "any",
+        "platform": "qemu",
+        "format": "qcow2",
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+        ],
+        "kargs": [
+            "console=tty0 console={terminal},115200n8",
+            "ignition.platform.id={platform}"
+        ],
+    },
+    "metal": {
+        "arch": "any",
+        "platform": "metal",
+        "format": "raw",
+        "kargs": [
+            "console=tty0 console={terminal},115200n8",
+            "ignition.platform.id={platform}"
+        ],
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+        ],
+    },
+    "dasd": {
+        "arch": "s390x",
+        "format":  "raw",
+        "platform": "metal",
+        "kargs": [
+            "ignition.platform.id={platform}"
+        ],
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+            "-device",
+            ("virtio-blk-ccw,drive=target,physical_block_size=4096,"
+             "logical_block_size=4096,scsi=off")
+        ],
+    }
+}
+
+
+ARCH_TERMINAL = {
+    "x86_64": "ttyS0",
+    "ppc64le": "hvc0",
+    "aarch64": "ttyAMA0",
+    "s390x": "ttysclp0",
+}
+
+
+def arch_terminal(arch):
+    """
+    Get the default terminal based on the architecture
+    """
+    if arch is None:
+        arch = get_basearch()
+    return ARCH_TERMINAL[arch]
+
+
+class MetalImageError(Exception):
+    """
+    Base error for build issues.
+    """
+    pass
+
+
+class MetalVariantImage(_Build, _BuildOSTree):
```

Comment:
I usually avoid multiple inheritance.  But, your design sense seems to drive to OO - I'm not opposed, I would just personally do it in a much less OO way.

And in this particular case it seems more right to say that a "metal image" *has-a* "ostreebuild" rather than *is-a* ostreebuild.  IOW we'd create an ostreebuild object inside here rather than inheriting.

I'd personally probably not have an "ostreebuild" object at all, just some freestanding helper functions in an `ostreebuild.py` module or so myself.

I guess part of all of this is to be honest I really don't like Python anymore :smile:

You don't need to change this, just my 2c.

Reply:
I thought about this some more and I think there's another way to say this:

To me coreos-assembler was something created because it was needed to accomplish building RHCOS, not because I wanted it as some end in itself.  I always thought of it more like a *script* - and scripts are more "procedural" than object oriented usually.  (They start and end, rather than running indefinitely, having other code call into and out of them library style).

And I personally don't much like Python anymore in and of itself - you'd never find me at PyCon or whatever. I just use it because it's fast to write and convenient to launch subprocesses.    I'm also very unlikely to ever write a new library for PyPy or whatever.

So a good example is I wouldn't go through all of the docstring types stuff and in particular the `@property` abstraction etc - it's just a script, it's OK if it's not "super beautiful Python" that you might find in a library.

But the perspective is probably different thinking of oneself as owning the project more itself, and obviously is going to differ for people who more *like* Python.

Anyways I'm happy to step back from this perspective and just accept the OO and "dense Python" style and try to spot bugs, this can be my last comment on this.

Reply:
Probably worth discussing this as a team during a cabal or something?

Reply:
Um, okay. I'll stand down. The OOO concerns have been heard.

Reply:
We did already talk about this before in https://github.com/coreos/coreos-assembler/pull/946

OK just one thing I also wanted to bring this up again is around [cosa+mantle](https://github.com/coreos/coreos-assembler/issues/163#issuecomment-574885334) - I'd still like to merge the projects and that brings the Python+Go thing straight front and center.  And again as I mentioned in #946 Go doesn't have inheritance - and since I write a lot of Go (along with Rust and C) I just don't tend to "think in objects".  Cleaning up code doesn't mean by default creating classes.

Also a data point: mantle is very "procedural" Go too in most places, not a lot of use of structs/interfaces.

This is overall still an improvement so from my PoV we can merge this and I will really try to stop commenting on OO again!

Reply:
I honestly do not know how to respond to this manifesto, but clearly @cgwalters you feel it important enough to have posted variations of it on three of my PRs.  Please take the topic to a cabal which is a more appropriate venue than my PR's.

Reply:
@cgwalters you have raised your objections to Pythonic and OO code thrice in my PR's.  Having these discussions in the narrow scope of a PR does not allow a fuller debate and muddies the pull-request. To simply accept your statement that you will "try to stop commenting on OO again" does not give your concerns voice. I am not interested in writing code that irritates other developers because its Pythonic. I would rather that we discuss it as a group, come to a decision and move on. Even if this PR never merges.

Reply:
I believe we did discuss this in a group in last years face to face. While it's fair to have preferences in programming style, I feel as if this thread is distracting from trying to make this project more supportable and hackable ... no matter the structures in use.

Reply:
It's not just about *this* project though - we maintain tools in a variety of languages.  And most crucially mantle and Go which interacts a lot with this project.

Reply:
If this PR is going to turn into a debate about the coding style, I'm done with it.

Reply:
Context matters. At the Face to Face, I proposed that consider using GoLang. However, as group, it was decided to pursue class-based Python to provide for code reuse and provide useful abstractions. I volunteered to pursue that because of its interesting work, and I wanted to see COSA become something more. The agreement and the goal have struck a nerve, apparently. 

This PR will not be re-opened.